sdk/metric/integrator/simple/simple.go

```diff
@@ -68,97 +84,209 @@ type (
+func New(aselector export.AggregationSelector, eselector export.ExportKindSelector) *Integrator {
+	now := time.Now()
 	return &Integrator{
-		AggregationSelector: selector,
-		stateful:            stateful,
-		batch: batch{
-			values:        map[batchKey]batchValue{},
-			intervalStart: time.Now(),
+		AggregationSelector: aselector,
+		ExportKindSelector:  eselector,
+		state: state{
+			values:        map[stateKey]*stateValue{},
+			processStart:  now,
+			intervalStart: now,
 		},
 	}
 }
 
-func (b *Integrator) Process(accumulation export.Accumulation) error {
+// Process implements export.Integrator.
+func (b *Integrator) Process(accum export.Accumulation) error {
 	if b.startedCollection != b.finishedCollection+1 {
 		return ErrInconsistentState
 	}
-
-	desc := accumulation.Descriptor()
-	key := batchKey{
+	desc := accum.Descriptor()
+	key := stateKey{
 		descriptor: desc,
-		distinct:   accumulation.Labels().Equivalent(),
-		resource:   accumulation.Resource().Equivalent(),
-	}
-	agg := accumulation.Aggregator()
-	value, ok := b.batch.values[key]
-	if ok {
-		// Note: The call to Merge here combines only
-		// identical accumulations.  It is required even for a
-		// stateless Integrator because such identical accumulations
-		// may arise in the Meter implementation due to race
-		// conditions.
-		return value.aggregator.Merge(agg, desc)
-	}
-	// If this integrator is stateful, create a copy of the
-	// Aggregator for long-term storage.  Otherwise the
-	// Meter implementation will checkpoint the aggregator
-	// again, overwriting the long-lived state.
-	if b.stateful {
-		tmp := agg
-		// Note: the call to AggregatorFor() followed by Merge
-		// is effectively a Clone() operation.
-		b.AggregatorFor(desc, &agg)
-		if err := agg.Merge(tmp, desc); err != nil {
-			return err
+		distinct:   accum.Labels().Equivalent(),
+		resource:   accum.Resource().Equivalent(),
+	}
+	agg := accum.Aggregator()
+
+	// Check if there is an existing record.
+	value, ok := b.state.values[key]
+	if !ok {
+		stateful := b.ExportKindFor(desc, agg.Aggregation().Kind()).MemoryRequired(desc.MetricKind())
+
+		newValue := &stateValue{
+			labels:   accum.Labels(),
+			resource: accum.Resource(),
+			updated:  b.state.finishedCollection,
+			stateful: stateful,
+			current:  agg,
+		}
+		if stateful {
+			// If stateful, allocate a cumulative aggregator.
+			b.AggregatorFor(desc, &newValue.cumulative)
+
+			if desc.MetricKind().PrecomputedSum() {
+				// If we need to compute deltas, allocate another aggregator.
+				b.AggregatorFor(desc, &newValue.delta)
+			}
 		}
+		b.state.values[key] = newValue
+		return nil
 	}
-	b.batch.values[key] = batchValue{
-		aggregator: agg,
-		labels:     accumulation.Labels(),
-		resource:   accumulation.Resource(),
+
+	// Advance the update sequence number:
+	sameRound := b.state.finishedCollection == value.updated
+	value.updated = b.state.finishedCollection
+
+	// An existing record will be found when:
```

Comment:
I don't really get this comment. It seems like an existing record will always be found on the second round since we always populate this map if its not set above. 

Is this supposed to say "An existing record will be found **in the same round** when:"? But then why would (a) effect this?

Reply:
I think this is confusing as well.

I think this is describing when an "existing record will _not_ be found in the same round" and `sameRound` will be false.

I think it would be helpful to be consistent in terminology here. Instead of `record` -> `value`, and instead of the term `round` maybe something more descriptive of the value's state. Maybe `continuous`?